libbpf-tools/klockstat.bpf.c

```diff
@@ -0,0 +1,265 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 Google LLC.
+ *
+ * Based on klockstat from BCC by Jiri Olsa and others
+ * 2021-10-26   Barret Rhoden   Created this.
+ */
+#include "vmlinux.h"
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "klockstat.h"
+#include "bits.bpf.h"
+
+const volatile pid_t targ_tgid = 0;
+const volatile pid_t targ_pid = 0;
+struct mutex *const volatile targ_lock = NULL;
+
+/* Make sure all programs are loaded before we start tracing */
+bool enabled = false;
```

Comment:
Why we need this variable ?

Reply:
there are a couple of mutexes that are grabbed while the three programs are inserted, and they lead to weird results.  i think the lock bookkeeping gets thrown off when we aren't tracing all three lock-entry/lock-exit/unlock in sync.  we'll think a task is at a certain depth, but not see the unlock.  later on, we see an unlock that we erroneously pair with an older lock.

specifically, this backtrace appears as if a mutex was held for the duration of klockstat.  this was taken from a 4 second duration.  the mutex wasn't actually held 4 seconds, but we thought it was.
```
                               Caller  Avg Hold    Count   Max Hold   Total Hold
           modify_ftrace_direct+0x18b 4002238347        1 4002238347   4002238347
          bpf_trampoline_update+0x427
       bpf_trampoline_link_prog+0x133
        bpf_tracing_prog_attach+0x26f
                      SYSC_bpf+0x2b3a
                              Max PID 2028, COMM klockstat
```

and if you periodically sample (with -i), you get some other bonkers results:

```
                               Caller  Avg Spin    Count   Max Spin   Total Spin
           modify_ftrace_direct+0x18b 6148913868173880962        3 18446741604521630976 18446741604521642888
          bpf_trampoline_update+0x427
       bpf_trampoline_link_prog+0x133
        bpf_tracing_prog_attach+0x26f
                      SYSC_bpf+0x2b3a
                              Max PID 2067, COMM klockstat
```

i'll add a comment to the source.